Ess adding in house functionalities

nitrosx opened the pull request:
THis pull adds to the official main branch in the SciCat project the changes that we have implemented in-house at ESS.

The code is still a work in progress and needs mroe testing.

nitrosx commented:
I forgot about the linting. Will work on it and update this PR

dylanmcreynolds commented on pyscicat/client.py:
```diff
@@ -86,6 +95,7 @@ def _send_to_scicat(self, url, dataDict=None, cmd="post"):
             response = requests.post(
                 url,
                 params={"access_token": self._token},
+                headers=self._headers,
```
I think that we should choose one or the other for sending the token, either as a query param or as a header. I can't remember why you preferred headers, but they are equal to me. So, can you remove the `access_token` query parameter?

nitrosx replied:
@dylanmcreynolds : current SciCat backend uses access_token, while the nestJs implementation is using headers. 
Given that soon we are migrating to the new one, and we, at ESS, are already testing the new implementation, I thought that it would be nice not to worry configuring pySciCat to work with a specific backend, but make it transparent. 
That's the reason why I left both of them.

That said, I understand that it would be less confusing to use only one method. 
Would you prefer to have a parameter when you instantiate the class that select which authentication model to use?
We could discuss at the next SciCat meeting

dylanmcreynolds replied:
Let's leave as you have done it in this PR for now. I have submitted an issue asking whether the nest.js backend should accept both methods here: https://github.com/SciCatProject/backend/issues/592

nitrosx replied:
I will check with Henrik, but I think nestjs implementation accept authentication only in the header, in the format:
`Authentication: Bearer <token>``

dylanmcreynolds replied:
Yes. I added an issue in backend to see if we can add the query parameter. There are probably numerous client code bases already that will be affected by this change.

dylanmcreynolds reviewed (changes requested):
Awesome stuff, thanks!

I added a couple of small comments.

Also, we should add unit tests for your new methods. 

https://github.com/SciCatProject/pyscicat/blob/main/pyscicat/tests/test_client.py

There's not a ton of logic in there, but the requests mock code can at least verify that we're pointing to the correct URL and test what logic is in there.

dylanmcreynolds commented:
Also, can you please look into adding documentation at https://github.com/SciCatProject/pyscicat/tree/main/docs/source? Maybe a couple of how to documents about getting proposals and instruments? I think the API calls you added will get auto generated, but we should check.

nitrosx commented on pyscicat/model.py:
```diff
@@ -198,6 +198,18 @@ class Datablock(Ownable):
     datasetId: str
 
 
+class OrigDatablock(Ownable):
```
Possibly!!!
I would say that we should follow what is done in SciCat backend for these two models.
I check and come back on this.

nitrosx replied:
@dylanmcreynolds I just checked on SciCat backend. The two models do not have any relation, but they do have some of the same fields. Possibly we can change the code and make Datablock inherit from origDatablock.
Thoughts?

dylanmcreynolds replied:
I assume that this this is an oversight on the server, but for only three fields I'm not going to worry about it, I think.

nitrosx replied:
I agree. I think we should change it first in the backend, than refactor the code here accordingly

nitrosx commented:
I will look into the documentation

nitrosx commented:
@dylanmcreynolds : I'm going to test pyScicat in a use case. I will have the examples to insert in the documentation once I'm done with that. I think it will take few days.
Do you want to wait for that to merge or should we go ahead merge and do another PR when I have the examples ready?

nitrosx commented:
@dylanmcreynolds : I updated the documentation as you requested.

nitrosx commented:
@dylanmcreynolds : is it ok if we merge this PR and I work and include testing in a new PR?

dylanmcreynolds commented on docs/source/howto/ingestion_simulation_dataset_ess.md:
```diff
@@ -0,0 +1,245 @@
+                "size": 448, 
+                "time": "2014-01-23T19:53:04.000Z"
+            },
+            ... omissed ...
```
In English, the word is "omitted"

dylanmcreynolds commented:
Thanks for the documentation!

The way it is right now, your hard work is not yet viewable in the documentation. 

Check out this https://github.com/SciCatProject/pyscicat/blob/b901bac087c7dde342589a58d3f4d689e350dcb3/docs/source/index.md?plain=1#L17

We could either add your documents as an and entry in "How To", or make a new category in the TOC called "Examples" for this documentation. What do you think?

BTW, this should be better documented, but to test the documentation build locally:

If you have not done so yet in your environment, setup the dev requirements to get the sphinx build code:
```
pip install -r requirements-dev.txt
```

Then:
```
cd docs
make html
```
This will build the web site locally into `docs/build/html`, and you can open a browser to the index.html

nitrosx commented:
@dylanmcreynolds : I added our example under the "How-to" section.
I fixed the json code in the document, so it builds.
I also added packages _pydantic_ and _requests_ in requirements-dev.txt, so to build the documentation the following instructions will suffice:
`
pip install -r requirements-dev.txt
cd docs
make html
`

dylanmcreynolds commented:
Sorry, my instructions assumed that you had already run: 

```
pip install -r requirements.txt
```
People tend treat requirements-dev.txt as "add-ons to the primary package". I'd rather not have dependencies repeeated in different files. Can you please remove this?

nitrosx commented:
No problem. Done!!!
We should include this info in the documentation. New ticket?

dylanmcreynolds commented:
> We should include this info in the documentation. New ticket?

Yes, good idea:
https://github.com/als-computing/pyscicat/issues/6
